### PR TITLE
New post: Node.js memory leak writeup

### PR DESCRIPTION
**Pull request by fstoerkle**
TODO:

- [ ] Add images to this repos and adjust image URLs
- [ ] Add link URLs

### CODE REVIEW

**Comment by wolframkriesing** [comment]
Cool insights, thanks for sharing!
If you like, I would find links to Stackdriver, npm modules and maybe some other things could be helpful. Otherwise there are some smaller wording things, that a polishing-read will uncover, sorry I didn't list them here.
Keep it up! Great Job 🚀🚀🚀

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 1]
```diff
@@ -0,0 +1,164 @@
+---
+layout: post
+title:  "Our journey discovering a Node.js memory leak"
+date:   2017-06-09 17:00:00 +0200
+categories: javascript node.js memoryleak bug
+author: Florian Störkle
+---
+
+# Tracking down a Node.js memory leak
+
```
Comment (gilligan) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: "Proved to be ..." or "turned out to be ..."

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+In this blog post we want to share our experience of tracking down and fixing a memory leak in one of our Node.js applications. For several reasons, this proved quite difficult, and took us several days to fix.
+
+## Preface
+
+Our team built a Node.js application that is responsible for synchronizing logs from an external service provider to our internal log pooling system (Google Stackdriver Logs). Let's call this application `sync-logs-to-stackdriver`.
+
+To make it simple, the job periodically checks if there are new logs available in the external system. If that’s the case it syncs them to Stackdriver. If not, it waits a short amount of time and then checks again for new logs.
```
Comment (gilligan) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: Guess that might just be personal preference but I think "that is" is nicer than "that's" here

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+In our internal container infrastructure, we have two instances of this service running:
+
+- a production instance (syncing ~ 100 logs/minute)
+- a development instance (syncing ~ 100 logs/day)
+
+
+## Discovery
+
+One day, we were noticed by our infrastructure team that one of applications, `sync-logs-to-stackdriver`,  killed and restarted automatically several times a day. They said the likely cause would be a memory leak.
```
Comment (gilligan) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: Noticed -> notified/informed/told.

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+As we are using NewRelic for monitoring our Node.js applications, we quickly checked the health of the application there. And indeed, we could confirm the suspicion: we really had a memory leak. This memory usage graph of the development instance of the application illustrates that:
+
+![NewRelic memory monitoring](https://github.hc.ag/pages/checkin/img/2017-05-10-memory-leak/dev-memory-leak-crashes.png)
+
+As one can see, the memory usage increases without upper border, until the container is killed by our infrastructure. Then the container is respawned, and it starts all over again.
+
+Interestingly, the leak was present in our production instance, but had far smaller impact – it only ran out of memory every other week.
+
+We noticed one thing that were quite suprised of that point in time: The memory leak was in the so-called "non-heap" memory of Node – that is the memory that is used by Node itself and binary extensions, but not by application code.
```
Comment (gilligan) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: Could you possibly turn non-heap into a link with more info?

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+main();
+```
+
+This made us even more confident that what we experienced is actually a memory leak of Node.js itself, and not caused by our application code.
+
+## The solution
+
+Having figured out the real problem, the solution was pretty simple: Node.js 7.10.0 has already been released, so we switched to using that for our application.
```
Comment (gilligan) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: has already been released -> *had* already been released.

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown (hunk 1 above)
Comment (fstoerkle) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: What do you mean with this suggestion? Adding it to the post's title?

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown (hunk 4 above)
Comment (fstoerkle) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: Hm unfortunately, I can't find any good resource on that.
Only [this article from NewRelic](https://blog.newrelic.com/2017/03/30/nodejs-vm-metrics-view-apm/) explains it the "New metrics, too" section:

> Non-Heap Memory used by Node for things like buffers and streams

Do you think that link would help nevertheless?

**Comment by fstoerkle** [comment]
Thanks all for your feedback, I address it so far. Will add images and more links during this week...

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+As we are using NewRelic for monitoring our Node.js applications, we quickly checked the health of the application there. And indeed, we could confirm the suspicion: we really had a memory leak. This memory usage graph of the development instance of the application illustrates that:
+
+![NewRelic memory monitoring](https://github.hc.ag/pages/checkin/img/2017-05-10-memory-leak/dev-memory-leak-crashes.png)
+
+As one can see, the memory usage increases without upper border, until the container is killed by our infrastructure. Then the container is respawned, and it starts all over again.
+
+Interestingly, the leak was present in our production instance, but had far smaller impact – it only ran out of memory every other week.
+
+We noticed one thing that were quite suprised of that point in time: The memory leak was in the so-called ["non-heap" memory](https://blog.newrelic.com/2017/03/30/nodejs-vm-metrics-view-apm/) of Node – that is the memory that is used by Node itself and binary extensions, but not by application code.
```
Comment (HeeL) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: `surprised`

File: _posts/2017-06-09-how-we-found-a-memory-leak.markdown [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+This made us even more confident that what we experienced is actually a memory leak of Node.js itself, and not caused by our application code.
+
+## The solution
+
+Having figured out the real problem, the solution was pretty simple: Node.js 7.10.0 had already been released, so we switched to using that for our application.
+
+After upgrading to the new Node.js version, we could no longer reproduce the memory leak. We were relieved 🎉🍻.
+
+Too be sure, we checked our NewRelic monitoring after some days. It clearly shows that we were able to fix the memory leak:
```
Comment (HeeL) [inline _posts/2017-06-09-how-we-found-a-memory-leak.markdown]: `To` be sure

**Comment by wolframkriesing** [comment]
please reopen this PR when it's ready for merging, currently its hanging here, and one always has to "filter" it out when one wants to merge a PR :) ... thanks

**Pull request closed without merging**